### PR TITLE
Backports 0.17 pr23

### PR DESCRIPTION
**Pull request by PastaPastaPasta**

### CODE REVIEW

File: src/evo/simplifiedmns.cpp [hunk 1]
```diff
@@ -187,19 +187,18 @@ bool BuildSimplifiedMNListDiff(const uint256& baseBlockHash, const uint256& bloc
 
     const CBlockIndex* baseBlockIndex = chainActive.Genesis();
     if (!baseBlockHash.IsNull()) {
-        auto it = mapBlockIndex.find(baseBlockHash);
-        if (it == mapBlockIndex.end()) {
+        baseBlockIndex = LookupBlockIndex(blockHash);
```
Comment (xdustinface) [inline src/evo/simplifiedmns.cpp:190]:
```suggestion
        baseBlockIndex = LookupBlockIndex(baseBlockHash);
```

Fixes `feature_dip4_coinbasemerkleroots` failure. `feature_fee_estimation` still fails but didn't get the time yet to look into it.

File: src/wallet/wallet.cpp [hunk 2]
```diff
@@ -5569,29 +5574,17 @@ int CMerkleTx::GetDepthInMainChain(const CBlockIndex* &pindexRet) const
     int nResult;
 
     if (hashUnset())
-        nResult = 0;
-    else {
-        AssertLockHeld(cs_main);
+        return 0;
```
Comment (UdjinM6) [inline src/wallet/wallet.cpp:5572]: changes below are from yet another PR, not even from 11041 it seems

File: src/wallet/rpcwallet.cpp [hunk 3]
```diff
@@ -3367,6 +3367,75 @@ UniValue fundrawtransaction(const JSONRPCRequest& request)
...
+
+    RPCTypeCheck(request.params, {UniValue::VSTR, UniValue::VARR, UniValue::VSTR}, true);
+
+    CMutableTransaction mtx;
+    if (!DecodeHexTx(mtx, request.params[0].get_str())) {
+        throw JSONRPCError(RPC_DESERIALIZATION_ERROR, "TX decode failed");
+    }
+
+    // Sign the transaction
+    LOCK2(cs_main, pwallet->cs_wallet);
```
Comment (UdjinM6) [inline src/wallet/rpcwallet.cpp:3435]:
```suggestion
    LOCK2(cs_main, mempool.cs);
    LOCK(pwallet->cs_wallet);
```
This should fix potential deadlocks.

File: test/functional/feature_dip3_deterministicmns.py [hunk 4]
```diff
@@ -444,7 +444,7 @@ def mine_double_spend(self, node, txins, target_address, use_mnmerkleroot_from_t
         amount -= Decimal("0.001") # fee
 
         rawtx = node.createrawtransaction(txins, {target_address: amount})
-        rawtx = node.signrawtransaction(rawtx)['hex']
+        rawtx = node.signrawtransactionwithwallet(rawtx)['hex']
```
Comment (UdjinM6) [inline test/functional/feature_dip3_deterministicmns.py:447]: `feature_dip0020_activation.py` needs a similar fix too

**Review by UdjinM6 — CHANGES_REQUESTED** [review]
9b3d61e should say smth like`Merge bitcoin#11041 Add LookupBlockIndex` maybe? Also it would be nice to have it rebased on top of develop to fix (some) test issues.

**Comment by PastaPastaPasta** [comment]
Rebased on develop and changed the name https://github.com/dashpay/dash/pull/3870/commits/df779026112ef6e7c290b3c6bdba97efba020413

File: test/functional/feature_dip3_deterministicmns.py (hunk 4 above)
Comment (PastaPastaPasta) [reply test/functional/feature_dip3_deterministicmns.py:447]: https://github.com/dashpay/dash/pull/3870/commits/54724d90740d0a93edccaeb3af5199621286cc20

File: src/wallet/wallet.cpp (hunk 2 above)
Comment (PastaPastaPasta) [reply src/wallet/wallet.cpp:5572]: Hmm, I agree these changes don't come from 11041... I can't find the PR it comes from... It looks like a good refactor. Can you find the PR it comes from?

**Comment by PastaPastaPasta** [comment]
Should've resolved all comments with these commits. Will squash after review

File: src/wallet/wallet.cpp (hunk 2 above)
Comment (UdjinM6) [reply src/wallet/wallet.cpp]: Found it: https://github.com/dashpay/dash/commit/f679e2431669d8fce270b2627b1908e1201f13e5#diff-1f2db0e4d5c12d109c7f0962333c245b49b696cb39ff432da048e9d6c08944d8R4056-R4086 tl;dr: it was modified to work with legacy IS "confirmations". This is no longer needed after https://github.com/dashpay/dash/commit/0a6f47323ba9da7a25e6bcb8b11aa4b21e6c71fd#diff-1f2db0e4d5c12d109c7f0962333c245b49b696cb39ff432da048e9d6c08944d8L5379-L5381, I think we can bring back the original logic here to reduce the amount of future merge conflicts. Could simply drop 6f1dbc1 OR squash it into df77902 and then add another commit on top which would revert Dash-specific code explicitly, I'm ok with either approach.

**Comment by UdjinM6** [comment]
Needs rebase + see https://github.com/dashpay/dash/pull/3870#discussion_r543194132

**Comment by PastaPastaPasta** [comment]
rebased, then dropped the revert commit

**Review by UdjinM6 — COMMENTED** [review]
utACK

**Review by UdjinM6 — DISMISSED** [review]
utACK

**Review by xdustinface — CHANGES_REQUESTED** [review]
Any reason why you didn't add the `LookupBlockIndex` in the remaining "pure dash" parts like in 5fb9f6d5f9f4808802521c2fda54a9474a1c273f?

**Comment by PastaPastaPasta** [comment]
I guess I was hoping you would do so :D

cherry picked and did a trivial "rebase" to place the commit in a good location

**Review by xdustinface — APPROVED** [review]
utACK

> I guess I was hoping you would do so :D

Haha, im pretty sure there was a commit related to the lookups with a commit message saying something like "in **most** dash code" when you opened the PR thats probably what made me think you left them out for a reason 😄

**Review by UdjinM6 — APPROVED** [review]
re-utACK

**Pull request merged**